src/main/java/com/arpnetworking/metrics/common/sources/KafkaSource.java

```diff
@@ -0,0 +1,222 @@
+        } catch (final InterruptedException e) {
+            LOGGER.warn()
+                    .setMessage("Unable to shutdown kafka consumer executor")
+                    .setThrowable(e)
+                    .log();
+        }
+    }
+
+    /**
+     * Generate a Steno log compatible representation.
+     *
+     * @return Steno log compatible representation.
+     */
+    @LogValue
+    public Object toLogValue() {
+        return LogValueMapFactory.builder(this)
+                .put("super", super.toLogValue())
+                .put("consumer", _consumer)
+                .build();
+    }
+
+    @Override
+    public String toString() {
+        return toLogValue().toString();
+    }
+
+    @SuppressWarnings("unused")
+    private KafkaSource(final Builder<T, V> builder) {
+        this(builder, LOGGER);
+    }
+
+    /* package private */ KafkaSource(final Builder<T, V> builder, final Logger logger) {
+        super(builder);
+        _logger = logger;
+        _consumer = builder._consumer;
+        _parser = builder._parser;
+        _runnableConsumer = new RunnableConsumerImpl.Builder<V>()
+                .setConsumer(builder._consumer)
+                .setListener(new LogConsumerListener())
+                .setPollTime(builder._pollTime)
+                .build();
+        _consumerExecutor = Executors.newSingleThreadExecutor(runnable -> new Thread(runnable, "KafkaConsumer"));
+    }
+
+    private class LogConsumerListener implements ConsumerListener<V> {
+
+        @Override
+        public void handle(final ConsumerRecord<?, V> consumerRecord) {
+            final T record;
+            try {
+                record = _parser.parse(consumerRecord.value());
+            } catch (final ParsingException e) {
+                _logger.error()
+                        .setMessage("Failed to parse data")
+                        .setThrowable(e)
+                        .log();
+                return;
+            }
+            KafkaSource.this.notify(record);
+        }
+
+        @Override
+        public void handle(final Throwable throwable) {
+            if (throwable instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+                _logger.info()
+                        .setMessage("Consumer thread interrupted")
+                        .addData("source", KafkaSource.this)
+                        .addData("action", "stopping")
+                        .setThrowable(throwable)
+                        .log();
+                _runnableConsumer.stop();
+            } else if (throwable instanceof KafkaException) {
+                _logger.error()
+                        .setMessage("Consumer received Kafka Exception")
+                        .addData("source", KafkaSource.this)
+                        .addData("action", "stopping")
+                        .setThrowable(throwable)
+                        .log();
+                _runnableConsumer.stop();
```

Comment:
Only throws errors for unrecoverable errors

Reply:
Can you link the docs here that state that? I would assume an unrecoverable error is typically (only?) encountered on start-up?

Can there be unrecoverable errors that are record specific (since this happens on read)? It seems problematic to kill the consumer in that event.

Specifically, the way the error handler is currently structured, a `ParsingException` will also trigger a stop. Right?

Reply:
https://kafka.apache.org/22/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#poll-java.time.Duration-

Docs for poll() method list all the exceptions that it throws. All the KafkaExceptions it throws don't seem recoverable except maybe WakeupException which I don't think should ever occur in our use case. These all seem like they would only happen on startup but the consumer doesn't actually "start-up" or connect to the cluster until poll is called.

A Parsing exception would be handled differently because a Kafka ConsumerRecord would be returned so the handle(ConsumerRecord) method would handle a parsing exception by just logging it and dropping that record.

We might not want to stop the consumer for other runtime exceptions caught in the else block but I followed the pattern in FileSource for that case.

Reply:
= Parsers & Consumers =

Consumer and ConsumerRecord are a Kafka types. Is the current plan to that each parser used by Kafka source implement its own Consumer? 

The way I see it there are two ways to allow re-use of existing parser implementations from Kafka:

1) Implement a Kafka Consumer that can have a configurable Parser. Then attach the Parser to a Kafka source via the standard consumer.
2) Implement a standard consumer that produces a byte-array and configure the Parser on the source.

I'm not sure what the pros and cons of each are. It should be possible to make both equally performant (wrt buffer copies). 

The former has the ability to NAK a request that it cannot parse (I believe) while the latter would make that more complex or impossible. That behavior could also not be varied by parser. I don't think this use case is extremely relevant at this point, as long as we have logs and metrics supporting any failure cases.

However, re-use of the existing parsers for data sourced from Kafka is a key integration requirement.

= Exceptions =

Yes, the FileSource shuts down on an interrupted exception.

https://github.com/ArpNetworking/metrics-aggregator-daemon/blob/master/src/main/java/com/arpnetworking/metrics/common/sources/FileSource.java#L200

However, on all other exceptions it logs, sleeps and continues. The exception may re-occur but it will be logged again and the code backs off. It is more likely to be obvious in the logs should this happen, then for the component to be quietly (it's not silent, but it's not yelling either) and the service to otherwise continue operating normally. 

= Metrics & Logging =

While not necessary on the first pass, I would like you to talk to Spencer about instrumenting the source with scoped or periodic metrics (FYI @speezepearson). This will be critical to successfully deploying and monitoring it.

Reply:
= Parsers & Consumers =
The current implementation is that the kafka source has a configurable KafkaConsumer and separate configurable Parser. (Case 2 above)
The plan is that the KafkaConsumer would use a standard deserializer to return a ConsumerRecord with a byte[]/bytebuffer value. This can then be passed to the parser to make a mad record. This will allow reuse of the other parsers because most take in a byte array or bytebuffer as input (which kafka already has deserializers for). I dont think there will be any issues with deserializing into a byte[] from kafka so we should always receive a ConsumerRecord. If the parser then cannot parse that array, the kafkasource will just log the parsing error and continue.
The only concern I see with this is that there is currently no validation that the deserialized value from kafka is the same input type as the parser.

= Exceptions =
I dont have strong feelings on this so I changed the handler to sleep and try again when an exception is received (even in "unrecoverable" cases), it will just continually log the error in those cases.

= Metrics & Logging =
TBD